### PR TITLE
[Identity] Make ClientAssertionCredential public

### PR DESCRIPTION
**Pull request by schaabs**
Adds `ClientAssertionCredential` to the public API surface
  - Adds classes `ClientAssertionCredential` and `ClientAssertionCredentialOptions`

### CODE REVIEW

File: sdk/identity/Azure.Identity/api/Azure.Identity.netstandard2.0.cs [hunk 1]
```diff
@@ -78,6 +78,17 @@ public partial class ChainedTokenCredential : Azure.Core.TokenCredential
         public override Azure.Core.AccessToken GetToken(Azure.Core.TokenRequestContext requestContext, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
         public override System.Threading.Tasks.ValueTask<Azure.Core.AccessToken> GetTokenAsync(Azure.Core.TokenRequestContext requestContext, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
     }
+    public partial class ClientAssertionCredential : Azure.Core.TokenCredential
+    {
+        public ClientAssertionCredential(string tenantId, string clientId, System.Func<string> assertionCallback, Azure.Identity.ClientAssertionCredentialOptions options = null) { }
```
Comment (xiangyan99) [inline sdk/identity/Azure.Identity/api/Azure.Identity.netstandard2.0.cs:83]: Does ClientAssertionCredential accept ResourceId?

Comment (schaabs) [reply sdk/identity/Azure.Identity/api/Azure.Identity.netstandard2.0.cs:83]: No, it must take a ClientId of an AAD application which is not a resource. Only customer assigned managed identities are resources and can be identified by their resource id.

**Review by JoshLove-msft — APPROVED** [review]

File: sdk/identity/Azure.Identity/src/ClientAssertionCredential.cs [hunk 2]
```diff
@@ -12,26 +12,57 @@
...
+            Client = options?.MsalClient ?? new MsalConfidentialClient(options?.Pipeline ?? CredentialPipeline.GetInstance(options), tenantId, clientId, assertionCallback, null, null, options?.IsLoggingPIIEnabled ?? false);
         }
 
-        public ClientAssertionCredential(string tenantId, string clientId, Func<string> getAssertionCallback, ClientAssertionCredentialOptions options = default)
+        /// <summary>
+        /// Creates an instance of the ClientCertificateCredential with the details needed to authenticate against Azure Active Directory using a signed client assertion.
+        /// </summary>
+        /// <param name="tenantId">The Azure Active Directory tenant (directory) Id of the service principal.</param>
+        /// <param name="clientId">The client (application) ID of the service principal</param>
+        /// <param name="assertionCallback">An asynchronous callback returning a valid client assertion used to authenticate the service principal.</param>
```
Comment (christothes) [inline sdk/identity/Azure.Identity/src/ClientAssertionCredential.cs:50]:
```suggestion
        /// <param name="assertionCallback">A synchronous callback returning a valid client assertion used to authenticate the service principal.</param>
```

File: sdk/identity/Azure.Identity/src/ClientAssertionCredential.cs [hunk 3]
```diff
@@ -12,26 +12,57 @@
...
+        public ClientAssertionCredential(string tenantId, string clientId, Func<CancellationToken, Task<string>> assertionCallback, ClientAssertionCredentialOptions options = default)
         {
+            TenantId = tenantId;
+            ClientId = clientId;
+            Client = options?.MsalClient ?? new MsalConfidentialClient(options?.Pipeline ?? CredentialPipeline.GetInstance(options), tenantId, clientId, assertionCallback, null, null, options?.IsLoggingPIIEnabled ?? false);
         }
 
-        public ClientAssertionCredential(string tenantId, string clientId, Func<string> getAssertionCallback, ClientAssertionCredentialOptions options = default)
+        /// <summary>
+        /// Creates an instance of the ClientCertificateCredential with the details needed to authenticate against Azure Active Directory using a signed client assertion.
```
Comment (christothes) [inline sdk/identity/Azure.Identity/src/ClientAssertionCredential.cs:46]: Should we call out more in the summary that the difference between these ctors is the sync/async callback?

File: sdk/identity/Azure.Identity/tests/ClientAssertionCredentialLiveTests.cs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+            // this assert is conditional because the access token is scrubbed in the recording so they will never be different
+            if (Mode != RecordedTestMode.Playback && Mode != RecordedTestMode.None)
+            {
+                Assert.AreNotEqual(token.Token, token2.Token);
+            }
+        }
+
+        [Test]
+        public async Task AuthnenticateWithAsyncAssertionCallback()
```
Comment (christothes) [inline sdk/identity/Azure.Identity/tests/ClientAssertionCredentialLiveTests.cs:67]: Could this be combined with the sync version of the test and take an argument to dedupe?

File: sdk/identity/Azure.Identity/tests/Azure.Identity.Tests.csproj [hunk 5]
```diff
@@ -24,4 +24,9 @@
     <ContentWithTargetPath Update="Data\*" CopyToOutputDirectory="PreserveNewest" TargetPath="certs\*" />
   </ItemGroup>
 
+  
+  <ItemGroup>
+    <Folder Include="SessionRecords\ClientAssertionCredentialLiveTests\" />
+  </ItemGroup>
+
```
Comment (christothes) [inline sdk/identity/Azure.Identity/tests/Azure.Identity.Tests.csproj:31]: delete?

**Review by christothes — APPROVED** [review]

**Pull request merged**